### PR TITLE
Add learning decay to the training

### PR DESCRIPTION
**Pull request by cutoken**
Add learning decay to the training:
1) It's constant decay per step for given number of steps
2) Taken through command line argument (this way in future you can add more arguments as needed)
3) Increase the default learning rate to 0.001 which is the high end of adam lr since it can now decay to the min_lr specified in the optimizer settings.

PS: Changed gitignore to ignore the train_data folder and Cargo.lock file

### CODE REVIEW

**Comment by keyvank** [comment]
How faster does this converge compared to the regular version?

**Comment by cutoken** [comment]
Current default learning rate is too low to converge fast. But if we increase it, the loss doesn't decrease after sometime. With decay, the loss drops fast initially and continues to drop. 1.8 loss is achievable for a 4L/4H/128ED model in around 5k iterations. After that it falls much slower (in line with default settings).

**Comment by keyvank** [comment]
@cutoken I wonder if the same behavior can be achieved by changing the `beta` values of Adam optimizer?

**Comment by cutoken** [comment]
That's a good idea @keyvank . What should the value be to simulate lr dropping from 0.001 to 0.00001 in 5k iterations (or may be iterations don't matter here ?)

**Comment by keyvank** [comment]
> That's a good idea @keyvank . What should the value be to simulate lr dropping from 0.001 to 0.00001 in 5k iterations (or may be iterations don't matter here ?)

I don't know. But interesting to try. We need dynamic learning-rates anyway. It's used in nanoGPT. So we will merge this after some refactorings. Btw, people say that GPT models need warm-up trainings, which means we should start with a very low learning-rate, go higher and higher until a point, and then make it fixed after some iters. (Exactly the reverse of what you have done haha)

https://stackoverflow.com/questions/55933867/what-does-learning-rate-warm-up-mean

**Comment by cutoken** [comment]
That's very interesting! I'll try the ramp up as well.
